module: ensure successful dynamic import returns the same result

aduh95 opened the pull request:
According to https://tc39.es/ecma262/#sec-HostLoadImportedModule, if `import()` is called multiple times with the same (`referrer`, `specifier`) pair, then it must fulfil return the same object (or reject).
https://tc39.es/proposal-import-attributes/#sec-HostLoadImportedModule takes it further by adding ~assertions~ attributes to the mix.

This PR adds a cache layer for dynamic imports to ensure Node.js meets the spec.
EDIT: I should clarify here that the cache layer is also populated by static imports

<!--
Before submitting a pull request, please read
https://github.com/nodejs/node/blob/HEAD/CONTRIBUTING.md.

Commit message formatting guidelines:
https://github.com/nodejs/node/blob/HEAD/doc/contributing/pull-requests.md#commit-message-guidelines

For code changes:
1. Include tests for any bug fixes or new features.
2. Update documentation if relevant.
3. Ensure that `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

GeoffreyBooth commented on lib/internal/modules/esm/loader.js:
```diff
@@ -71,6 +74,8 @@ class ESMLoader {
   #defaultLoad;
   #importMetaInitializer;
 
+  #dynamicImportCache = new SafeMap();
```
Why would we create a new cache for dynamic import specifically? Shouldn’t these already be in the module map?

Shouldn’t static and dynamic import of the same input return the same result? Wouldn’t that mean they need to share the same cache?

ljharb replied:
Indeed, dynamic and static imports must share the same cache.

GeoffreyBooth reviewed (changes requested):
I have two issues with this:
1. It's unclear what problem this solves. Is there a bug report? A spec violation that can be demonstrated in a repro?
2. We already have a module cache: `ModuleMap`. It feels wrong to create a second one. If the first one behaves incorrectly, it should be fixed or replaced. Having two seems likely to cause problems.

devsnek commented:
Node's module map should be a stricter version of the one specified in ecma262, specifically because we resolve the canonical path for a module, which means it isn't just stable within each file, but across the entire graph of files.

aduh95 commented:
> 1. Is there […] a spec violation that can be demonstrated in a repro?

Yes, see the test I've added

> 2\. We already have a module cache: `ModuleMap`. It feels wrong to create a second one. If the first one behaves incorrectly, it should be fixed or replaced

Having two seems necessary, there are simply two different things, the first one is behaving correctly AFAICT.

jkrems commented:
> According to https://tc39.es/ecma262/#sec-HostLoadImportedModule, if import() is called multiple times with the same (referrer, specifier) pair, then it must fulfil return the same object (or reject).

Isn't the same thing true for `import('x')` and `import 'x'` from the same referrer? Why do we need a secondary cache for import specifically unless that invariant is broken right now?

aduh95 commented:
> Isn't the same thing true for `import('x')` and `import 'x'` from the same referrer?

The same thing is true yes, this PR ensures that both static and dynamic imports returns the same thing.

GeoffreyBooth commented:
Migrating https://github.com/nodejs/node/pull/46826#issuecomment-1573748223 over here:

> ```js
> const importVal1 = await import('module-name');
> 
> setTimeout(() => {
>   const importVal2 = await import('module-name');
> 
>   console.log(importVal1 === importVal2);
> }, 999);
> ```
> 
> The ECMAScript spec enforces that the above code always logs true

A simplified version of the above is this:

```js
import('data:text/javascript,export default 333').then(async first => {
  const second = await import('data:text/javascript,export default 333')
  console.log(first === second)
})
```

In Node 20 for me, this logs `true`. If you replace the data URIs with `./file.mjs` and create a `file.mjs` containing `export default 333`, it still returns true. So it would seem that we’re already spec compliant, at least for the common use case where there are no I/O issues.

I looked at the test in this PR, and it involves filesystem mutations. Please correct me if I’m misreading this, but it seems like what you’re testing is an initial `import()` succeeding, then deleting the imported file, and the second import currently fails because the file is missing but it should have succeeded because the initial success should have been cached? Is that the intent of the test, and the change that this PR is intending to make?

jkrems commented:
> The same thing is true yes, this PR ensures that both static and dynamic imports returns the same thing.

Why isn't the `ModuleMap` cache ensuring that? If it doesn't, we should delete it so we have one cache that works instead of one that works and one that doesn't. Is the problem that the ModuleMap doesn't contain entries for failed loads? If so, can we add that to the module map cache?

GeoffreyBooth commented:
> Is the problem that the ModuleMap doesn’t contain entries for failed loads? If so, can we add that to the module map cache?

Per the spec we’re not supposed to caching failures, only successes. So if you tried to import a nonexistent file, then created the file and tried again to import it, the second attempt should succeed. https://tc39.es/ecma262/#sec-HostLoadImportedModule (emphasis added):

> If this operation is called multiple times with the same (referrer, specifier) pair and it performs [FinishLoadingImportedModule](https://tc39.es/ecma262/#sec-FinishLoadingImportedModule)(referrer, specifier, payload, result) **where result is a [normal completion](https://tc39.es/ecma262/#sec-completion-record-specification-type),** then it must perform [FinishLoadingImportedModule](https://tc39.es/ecma262/#sec-FinishLoadingImportedModule)(referrer, specifier, payload, result) with the same result each time.

But I also feel like this should be achievable with our current `ModuleMap` cache without needing to create a new cache. Is the issue that we’re checking the filesystem for every `import` before looking to see if a module is already in our cache?

aduh95 commented:
> But I also feel like this should be achievable with our current `ModuleMap` cache without needing to create a new cache. Is the issue that we’re checking the filesystem for every `import` before looking to see if a module is already in our cache?

Yes that's the issue, `defaultResolve` is making fs operation even if a previous resolution has succeeded in the same module. I don't think we have another option than caching all the successful resolve operations.

GeoffreyBooth commented:
> Yes that’s the issue, `defaultResolve` is making fs operation even if a previous resolution has succeeded in the same module.

Where is `defaultResolve` making filesystem calls? I read through it and I’m not seeing them.

Perhaps it was doing so before #47824?

aduh95 commented:
Of course it's doing fs operation, it needs to in order to resolve symlinks and node_modules deps.

Here's a few examples, `defaultResolve` is calling `moduleResolve` here:

https://github.com/nodejs/node/blob/ccada8bccc3ec90fa895b9f19ae37a460c318f60/lib/internal/modules/esm/resolve.js#L1035-L1040

Which it self is calling `packageResolve` and `finalizeResolution`:

https://github.com/nodejs/node/blob/ccada8bccc3ec90fa895b9f19ae37a460c318f60/lib/internal/modules/esm/resolve.js#L830
https://github.com/nodejs/node/blob/ccada8bccc3ec90fa895b9f19ae37a460c318f60/lib/internal/modules/esm/resolve.js#L837

And here where they make the FS call:

https://github.com/nodejs/node/blob/ccada8bccc3ec90fa895b9f19ae37a460c318f60/lib/internal/modules/esm/resolve.js#L211-L212
https://github.com/nodejs/node/blob/ccada8bccc3ec90fa895b9f19ae37a460c318f60/lib/internal/modules/esm/resolve.js#L749-L750

GeoffreyBooth commented:
Within the error path of one of these failed filesystem calls, could we look in `ModuleMap` to see if the requested module was cached? That way we would avoid both creating a new cache and slowing down resolution in the default case where the filesystem hasn’t changed under our feet.

I’m also not convinced that this is a real problem. What is the use case where `import` calls need to continue working correctly after files were deleted from disk? Even if that’s technically the correct behavior per spec, it’s such an extreme edge case that it doesn’t feel like something we should add a performance regression to achieve.

bmeck commented:
Beware. Read up on why the web does cache errors. Spec does not mandate it
across import sites but within a single importing source text it must keep
the error. Even still cross url errors are cached and multiple requests to
change that behavior haven't changed their behavior. Remember, once it
isn't an error the behavior will be cemented and very hard to change so
minor edge cases need consideration

On Fri, Jun 2, 2023, 3:51 PM Geoffrey Booth ***@***.***>
wrote:

> Within the error path of one of these failed filesystem calls, could we
> look in ModuleMap to see if the requested module was cached? That way we
> would avoid both creating a new cache and slowing down resolution in the
> default case where the filesystem hasn’t changed under our feet.
>
> I’m also not convinced that this is a real problem. What is the use case
> where import calls need to continue working correctly after files were
> deleted from disk? Even if that’s technically the correct behavior per
> spec, it’s such an extreme edge case that it doesn’t feel like something we
> should add a performance regression to achieve.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/nodejs/node/pull/46662#issuecomment-1574300171>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABZJIYX3IJLSRSKF47FALDXJJG3XANCNFSM6AAAAAAU4F7M7E>
> .
> You are receiving this because you are on a team that was mentioned.Message
> ID: ***@***.***>
>

GeoffreyBooth commented:
> Beware. Read up on why the web does cache errors.

I think what we’re discussing is the opposite though, caching the success so that subsequent imports continue to succeed even if the referenced files have been deleted.

aduh95 commented:
> I’m also not convinced that this is a real problem. What is the use case where `import` calls need to continue working correctly after files were deleted from disk?

I think that's off topic: if you don't like the spec, you should take it to TC39, but until then it seems pointless to discuss this in this PR.



> Within the error path of one of these failed filesystem calls, could we look in `ModuleMap` to see if the requested module was cached?

I think it's still a problem if the resolution ends up on a different symlink (so we never end up on the error path yet we are breaking the spec).

> Read up on why the web does cache errors

Caching the errors would simplify the implementation, it's certainly something we can think about.

devsnek commented:
I haven't read the code of this PR yet but I'm generally in favor of the behavior the spec requires here. We spent a great deal of time coming to the exact semantics we wanted to preserve and anyone who feels differently may want to review our prior discussions on this (you can check the dynamic import proposal repo and tc39 notes from that time) before pushing too hard.

GeoffreyBooth commented on test/es-module/test-esm-dynamic-import-mutating-fs.js:
```diff
@@ -0,0 +1,25 @@
+
+  await fs.writeFile(target, 'export default "actual target"\n');
+
+  const moduleRecord = await import(target);
+
+  await fs.rm(target);
+
+  assert.strictEqual(await import(target), moduleRecord);
```
To ensure that we’re comparing against what was exported, and not inadvertently against an empty object or an error object or something else.

```suggestion
  await import(target);

  await fs.rm(target);

  assert.strictEqual(await import(target), 'actual target');
```

aduh95 replied:
No! The spec says:

> If this operation is called multiple times with the same (referrer, specifier) pair and it performs FinishLoadingImportedModule(referrer, specifier, payload, result) where result is a normal completion, then it must perform FinishLoadingImportedModule(referrer, specifier, payload, result) with the same result each time.

So it doesn't matter what the result is, what we are trying to validate is whether the dynamic import returns the same object each time. It doesn't matter for this test if it's an empty object, or an error, or whatever: the only thing we care is wether it's "the same result". Other tests ensures that dynamic imports resolve to the correct value, I don't see any value in adding that assertion to this test.

GeoffreyBooth commented on test/es-module/test-esm-dynamic-import-mutating-fs.mjs:
```diff
@@ -0,0 +1,42 @@
+
+await fs.writeFile(target, 'export default "actual target"\n');
+
+const moduleRecord = await import(target);
+
+await fs.rm(target);
+
+assert.strictEqual(await import(target), moduleRecord);
```
```suggestion
await import(target);

await fs.rm(target);

assert.strictEqual(await import(target), 'actual target');
```

GeoffreyBooth commented on test/es-module/test-esm-dynamic-import-mutating-fs.mjs:
```diff
@@ -0,0 +1,42 @@
+  await spawnPromisified(execPath, [
+    '--input-type=module',
+    '--eval',
+    [`import * as d from${JSON.stringify(target)};`,
+     'import{rm}from"node:fs/promises";',
+     `await rm(new URL(${JSON.stringify(target)}));`,
+     'import{strictEqual}from"node:assert";',
+     `strictEqual(JSON.stringify(await import(${JSON.stringify(target)})),JSON.stringify(d));`].join(''),
```
```suggestion
    `
    import ${JSON.stringify(target)};
    import { rm } from "node:fs/promises";
    await rm(new URL(${JSON.stringify(target)}));
    import { strictEqual } from "node:assert";
    strictEqual(JSON.stringify(await import(${JSON.stringify(target)})), "actual target");`,
```

GeoffreyBooth commented on test/es-module/test-esm-initialization.mjs:
```diff
@@ -8,22 +8,22 @@ import { describe, it } from 'node:test';
 describe('ESM: ensure initialization happens only once', { concurrency: true }, () => {
   it(async () => {
     const { code, stderr, stdout } = await spawnPromisified(execPath, [
+      '--experimental-import-meta-resolve',
```
Why is this being added? I thought we removed this flag, or were going to do so immediately after landing off-thread.

aduh95 replied:
I think it was forgotten when the test was added, `runmain.js` use `import.meta.resolve`.

GeoffreyBooth replied:
Okay so this is just an unrelated change? Should this be in its own PR?

GeoffreyBooth commented on test/es-module/test-esm-initialization.mjs:
```diff
@@ -8,22 +8,22 @@ import { describe, it } from 'node:test';
+     * 4. doesnt-matter.mjs (second import.meta.resolve call)
      */
-    assert.strictEqual(resolveHookRunCount, 2);
+    assert.strictEqual(stdout.match(/resolve passthru/g)?.length, 4);
```
This feels like a problem? Resolutions are happening twice as many times now as before?

aduh95 replied:
See the comments above, it's related to `import.meta.resolve` being called twice in the test.

GeoffreyBooth replied:
But you didn’t change this test, just the assertions for it to pass. Doesn’t that mean that the changes in this PR break the previous test? Why should this test need to change based on cache refactoring?

Or was the test just buggy before because it was missing the required `--experimental-import-meta-resolve` flag, and now that it has the flag it should’ve had all along, the assertions need to change accordingly? If so, this feels like an unrelated fix that should be its own PR.

GeoffreyBooth commented on lib/internal/modules/esm/loader.js:
```diff
@@ -237,13 +251,60 @@ class DefaultModuleLoader {
+      return specifierCache[serializedAttributes];
+    }
+
+    // If it's still a promise, we must have a fallback in case it fails:
+    const fallback = () => {
+      // If another fallback has already cached a promise, use this one:
+      if (specifierCache[serializedAttributes] != null) {
+        return PromisePrototypeThen(specifierCache[serializedAttributes], undefined, fallback);
+      }
+      // Otherwise create a new cache entry:
+      const result = this.#import(specifier, parentURL, importAssertions);
+      specifierCache[serializedAttributes] = result;
+      PromisePrototypeThen(result, undefined, removeCache);
+      return result;
+    };
+    return PromisePrototypeThen(specifierCache[serializedAttributes], undefined, fallback);
```
Do we have test coverage for all of this logic?

aduh95 replied:
The test cases I added.

GeoffreyBooth reviewed (commented):
@aduh95 and I discussed this. We came to the conclusion that as far as we were aware, our ESM loader implementation was caching the results of the entire module resolution and loading process but not the individual results of the resolve and load steps. This is a problem for the spec compliance issue that Antoine pointed out, where a resolution success needs to be cached even if subsequent attempts at resolving the same input fail; without a resolution cache, we have no way of skipping repeated attempts at resolving the same input and therefore adhering to the spec.

I suggested therefore creating two caches: a “resolve cache” and a “load cache,” to correspond with the two phases/hooks of module resolution and loading. This should both fix the spec compliance issue as well as improve performance whenever an application imports the same thing twice (which is probably rare, as _all_ of the input would need to be identical including parent URL, and it’s uncommon to have two `import` statements in the same file for the same thing). This PR creates the new “resolve cache” and renames the former “module map” to be the “load cache.”

One thing that bothered me about the former module map, and is still problematic in this version, is that the values of the module map/load cache are sometimes promises, sometimes “module jobs,” etc rather than a one-to-one mapping of the output of the `load` hook. And in this PR, the value of the resolve cache is the same promise or module job as the value of the module map/load cache. I would think that the value of the resolve cache should be whatever gets returned by `resolve`: the absolute URL of the module and its assertions/attributes. Then that value would be the key for the load cache. 

I see the sense in making the value of the resolve cache the same as the value of the load cache, to skip the second lookup and therefore save a tiny bit of performance. It just breaks the mental model of the purposes of these two caches, if the resolve cache isn’t really caching the results of the `resolve` step. And the load cache isn’t really caching the result of the `load` step either, as a module job is a much larger object than simply the module source that gets returned from `load`.

Is what we really want here a single big cache where module jobs (or the promises that will resolve into module jobs) can be looked up by two keys, either the input for `resolve` or the input for `load`? Is that the most performant data structure for what we’re trying to achieve here?

For me what makes the most sense mentally in terms of my ease in following the logic would be something like this:
- A “resolve cache” where the keys are the serialized inputs to `resolve` and the values are the successful results of those `resolve` calls (and failures aren’t cached)
- A “load cache” where the keys are the serialized inputs to `load` and the values are the successful results of those `load` calls (and failures aren’t cached)

And do we even need a data structure to hold the pending promises of the “module jobs,” the process of loading the module sources from disk or network? I suppose we probably do, to avoid parallel `load` calls trying to read the same file from disk unnecessarily, which would mean yet a third cache where the keys are the same as those of the load cache and the values are the promises, or a boolean to tell you that it’s still pending, or something. 

Just writing this all out it makes me feel like creating all these extra data structures would be bad for performance, so I’m not trying to say that I’m advocating for this architecture; just that it feels like what would make the most sense in terms of me being able to follow the logic as a reader. Is there some version of refactoring the module map that we have, with creating the resolution cache, that comes a bit closer to this mental model?

aduh95 commented on lib/internal/modules/esm/loader.js:
```diff
@@ -16,14 +16,20 @@ const {
-function newModuleMap() {
-  const ModuleMap = require('internal/modules/esm/module_map');
-  return new ModuleMap();
+function newModuleResolveMap() {
+  const { ModuleResolveMap } = require('internal/modules/esm/module_map');
+  return new ModuleResolveMap();
+}
+
+function newModuleLoadMap() {
+  const { ModuleLoadMap } = require('internal/modules/esm/module_map');
+  return new ModuleLoadMap();
 }
```
We call them in the tests, and yes also when we have several loaders (although now I don't think it ever happens on the same thread, so that last one would not be a concern).

aduh95 commented on lib/internal/modules/esm/module_job.js:
```diff
@@ -72,10 +72,12 @@ class ModuleJob {
         ArrayPrototypePush(dependencyJobs, job);
-        return job.modulePromise;
+        const result = await job.modulePromise;
+        this.loader.cacheStaticImportResult(specifier, url, attributes, job);
```
We don't have access to `this.#resolveCache` here, it's a different class – and I figured it was better not to expose it.

aduh95 commented on lib/internal/modules/esm/module_map.js:
```diff
@@ -1,17 +1,64 @@
+        (key) => JSONStringify(key) + JSONStringify(importAssertions[key])),
+      ',');
+
+    return { specifierCache, serializedAttributes };
```
I use it either to set it to a new value (`specifierCache[serializedAttributes] = …`) or to get the cached value (`specifierCache[serializedAttributes]`), and sometime both. It's not creating any placeholder entry no.

aduh95 commented on lib/internal/modules/esm/loader.js:
```diff
@@ -237,13 +251,60 @@ class DefaultModuleLoader {
    * @param {string} parentURL Path of the parent importing the module.
    * @param {Record<string, string>} importAssertions Validations for the
    *                                                  module import.
-   * @returns {Promise<ExportedHooks | KeyedExports[]>}
```
For the sake of simplifying my life, can you include the added line as well for this kind of comments? It's a bit hard to comment without having the context and for some reason the GitHub UI only shows the line above.

The JSDoc is wrong/outdated, loader.import never returns an array – it used to before https://github.com/nodejs/node/pull/48296 and I guess we forgot about JSDoc in that PR. `ExportedHooks` doesn't seem to fit very well here either, if you prefer we can address that in its own PR.

JakobJingleheimer replied:
I think `ModuleExports` is correct (`import` is not limited to hooks).

I think `KeyedExports` is obsolete as of https://github.com/nodejs/node/pull/48296: import previously returned `{ [moduleUrl]: ModuleExports }` in order for the caller to know what exports belonged to which module (because multiple were possible).

JakobJingleheimer reviewed (commented):
In terms of unsettled promises, could we perhaps initially set the value to the pending promise, and then upon settling, overwrite the promise with its settled value?

aduh95 commented:
> In terms of unsettled promises, could we perhaps initially set the value to the pending promise, and then upon settling, overwrite the promise with its settled value?

Unless I misunderstand what you mean, I think that's what this PR is doing already.

aduh95 commented on lib/internal/modules/esm/loader.js:
```diff
@@ -237,13 +251,60 @@ class DefaultModuleLoader {
    * @param {Record<string, string>} importAssertions Validations for the
    *                                                  module import.
-   * @returns {Promise<ExportedHooks | KeyedExports[]>}
+   * @returns {Promise<ExportedHooks>}
```
Good catch, I'm going to take that to its own PR: https://github.com/nodejs/node/pull/48424

aduh95 commented on lib/internal/modules/esm/loader.js:
```diff
@@ -133,8 +144,7 @@ class DefaultModuleLoader {
    *                                     point.
-   * @param {Record<string, string>} importAssertions Validations for the
-   *                                                  module import.
+   * @param {Record<string, string>} importAssertions The import attributes.
```
```suggestion
   * @param {Record<string, string>} importAssertions Validations for the
   *                                                  module import.
```

GeoffreyBooth commented:
So this PR is essentially a significant refactor of `module_map.js`. We also have https://github.com/nodejs/node/pull/48439 open that’s _also_ a significant refactor of `module_map.js`, though with different goals. We should probably consolidate the two somehow, so that a single PR both untangles the module map logic that @JakobJingleheimer and I find convoluted, while at the same time fixing the bug that this PR fixes.

Currently https://github.com/nodejs/node/pull/48439 is paused because https://github.com/nodejs/node/pull/48559 needs to land first. I think the next steps are:

- Get https://github.com/nodejs/node/pull/48559 over the finish line and landed.
- @JakobJingleheimer and @aduh95 review their two PRs and agree on an approach: one PR or two, who’s doing what, and if two PRs which one lands first.

What do you think?

aduh95 commented:
https://github.com/nodejs/node/pull/48439 unflags the loader implementation, I'd argue that we want that to land last.

GeoffreyBooth commented:
> #48439 unflags the loader implementation, I’d argue that we want that to land last.

https://github.com/nodejs/node/pull/48559 also unflags it I think. See https://github.com/nodejs/node/pull/48559#discussion_r1260335580.

Currently the `register` PR is landed but marked as don’t backport, so all these subsequent PRs that build on it could be the same. We could just keep all the landed PRs as unreleased until we land enough to feel comfortable unflagging, and then release everything that’s landed by that point. The Loaders API would remain experimental and still print the experimental warning, but it would move into a new experimental phase that indicates that it’s getting close to stability.

So what this looks like, I think, is that we should land all the PRs that complete [Milestone 2](https://github.com/nodejs/loaders#milestone-2-stability) and then release them all at once, which would unflag the Loaders API but leave it as experimental until we consider it stable. That would mean landing the following:

- https://github.com/nodejs/node/pull/48559
- The PR that comes out of https://github.com/nodejs/loaders/issues/147 (remove `globalPreload`, move its functionality into `register` / `initialize`); there are already branches for this but not yet opened as a PR.
- https://github.com/nodejs/node/pull/47999

Once those three are all landed, we can untag the `dont-backport` label and release all of them, and the implementation becomes unflagged. Then this PR and/or whatever https://github.com/nodejs/node/pull/48439 becomes can also land.

We can also discuss in https://github.com/nodejs/loaders/issues/149.

JakobJingleheimer commented:
We're gonna go for https://github.com/nodejs/node/pull/48559, so I think the only thing to keep from my PR (https://github.com/nodejs/node/pull/48439) is the module_map untangling. I think the best way to "save" that work from my PR is if Antoine is able to incorporate it into his here (I'm happy to help, but this PR is his brainchild, so I think he's better suited to do it), or, if that's not really viable, then we try to apply my stuff after this lands.

GeoffreyBooth commented on lib/internal/modules/esm/module_map.js:
```diff
@@ -1,17 +1,119 @@
+   */
+  getSerialized(specifier, parentURL, importAssertions) {
+    let internalCache = super.get(parentURL);
+    if (internalCache == null) {
+      super.set(parentURL, internalCache = { __proto__: null });
+    }
```
What is `internalCache`? It seems like _another_ key-value map, but one for each `parentURL`? As in, these are all the imports from each module? Why is this a plain object instead of another `SafeMap`?

aduh95 replied:
I copied the implementation from `LoadModuleCache`. There might be performance considerations, or the original author found it more readable this way, not sure.

GeoffreyBooth replied:
We seem to use `SafeMap` rather than plain objects for dictionaries throughout the codebase, so I'd guess that `SafeMap` is more performant?

JakobJingleheimer replied:
`Map` is HUGELY more performant over a plain object for a large number of rapid mutations

GeoffreyBooth commented on lib/internal/modules/esm/module_map.js:
```diff
@@ -1,17 +1,119 @@
+      super.set(parentURL, internalCache = { __proto__: null });
+    }
+
+    // To serialize the ModuleRequest (specifier + list of import attributes),
+    // we need to sort the attributes by key, then stringifying,
+    // so that different import statements with the same attributes are always treated
+    // as identical.
```
Are we doing anything like this serialization today? If not, wouldn’t adding it be a big performance hit? Can you run any benchmarks to see how much of an impact this has?

I’m still ambivalent as to whether this fix needs to happen at all, seeing as we’ve had no user complaints about it, but if it also incurs a noticeable performance hit then I’d be more inclined to go back to TC39 and ask that the spec be updated to make the behavior host-defined (so if browsers want to return the cached result, fine, but Node can choose otherwise).

aduh95 replied:
You're welcome to run the benchmark CI and/or complain to TC39 (but consider https://github.com/nodejs/node/pull/46662#issuecomment-1574452202 first), but I'm not sure I understand what's the point, we want to be spec-compliant before being performant.

GeoffreyBooth replied:
We're out of compliance today and there isn't a rush to get into compliance. If we contact them and a change is possible, we could just stay as we are and wait for them to update.

GeoffreyBooth commented on lib/internal/modules/esm/loader.js:
```diff
@@ -225,8 +240,50 @@ class DefaultModuleLoader {
+    const { internalCache, serializedModuleRequest } =
+      this.#resolveCache.getSerialized(specifier, parentURL, importAssertions);
+    const removeCache = () => {
+      // Remove the cache entry if the import fails.
+      delete internalCache[serializedModuleRequest];
```
This only removes the entry _relative to its parent,_ right? Since `internalCache` is defined for each `parentURL`? Shouldn’t failures remove the cache entry everywhere? Is there a way to define the resolve cache such that it’s not a map of maps, but closer to a straightforward key-value cache?

aduh95 replied:
> Shouldn’t failures remove the cache entry everywhere?

No, if `import("./file.js")` fails in `a/entry.js`, that shouldn't invalidate the cache entry for `import("./file.js")` in `b/other_file.js`.

GeoffreyBooth replied:
Okay this makes sense. Perhaps it should be named something like `parentURLImports` to be clearer what it is? And like all the caches, a comment explaining the keys and values would help.

GeoffreyBooth commented on test/es-module/test-esm-loader-modulemap.js:
```diff
@@ -24,11 +24,11 @@ const jsonModuleJob = new ModuleJob(loader, stubJsonModule.module,
 
-// ModuleMap.set and ModuleMap.get store and retrieve module jobs for a
-// specified url/type tuple; ModuleMap.has correctly reports whether such jobs
+// LoadCache.set and LoadCache.get store and retrieve module jobs for a
+// specified url/type tuple; LoadCache.has correctly reports whether such jobs
 // are stored in the map.
```
Assuming we keep the resolve cache (rather than collapse them back into one cache per my other comment) then shouldn’t we have similar tests for `ResolveCache.set` and `setLazy` and `getSerialized` and so on?

GeoffreyBooth reviewed (commented):
Mostly just questions, that maybe imply that we should add a few more explanatory comments in here. The main thing that bothers me is that I don’t know what the `resolve` cache is: I would assume based on the name that it’s a key-value map where the key is the `resolve` input and the value is the `resolve` output, but as far as I can tell that’s not what it really is here (and the value seems to be a map of maps, adding more complexity). If it’s more efficient to just keep it as one map, I’m fine with it going back to `moduleMap` and we can add additional keys so that we can retrieve cached results based on either `resolve` input or `resolve` output, assuming that that’s the most efficient data structure; or we can keep the current design or something close to it if you think that that’s what makes the most sense. Whatever we do, I just want a detailed comment explaining each cache’s keys and values (and if the value is itself a map, then _that_ map’s keys and values) so that I can make sense of this. I find this part of the codebase very confusing.

Also if the additional serialization is something new and something that we need to run for every import, we should run some benchmarks to see the performance impact. I would think the impact should be very minimal if we can avoid serialization for `import`s that lack assertions (see one of my notes).

aduh95 commented on lib/internal/modules/esm/module_map.js:
```diff
@@ -1,17 +1,119 @@
+    // we need to sort the attributes by key, then stringifying,
+    // so that different import statements with the same attributes are always treated
+    // as identical.
+    const serializedModuleRequest = specifier + ArrayPrototypeJoin(
+      ArrayPrototypeMap(
+        ArrayPrototypeSort(ObjectKeys(importAssertions)),
+        (key) => JSONStringify(key) + JSONStringify(importAssertions[key])),
+      ',');
```
They should never be undefined, I like the idea of a shortcut for the empty case but I don't see we could detect that.

GeoffreyBooth replied:
Isn't it set to `kEmptyObject` by default? So compare it to that?

aduh95 replied:
It depends what you mean by "by default", if the user writes `import("some-specifier")` (i.e. a dynamic import without any import attribute), we won't get `kEmptyObject` (we'll get a new empty null-prototype object). The case where we'd actually get `kEmptyResult` is so rare that it seems pointless to optimize for.

aduh95 commented on lib/internal/modules/esm/module_map.js:
```diff
@@ -1,17 +1,119 @@
+   * @param {Record<string, string>} importAttributes
+   * @returns {import('./loader').ModuleExports | Promise<import('./loader').ModuleExports>}
+   */
+  get(specifier, parentURL, importAttributes) {
```
I'm using the updated spec though, this is following what import attributes are, not import assertions. The inconsistency is regretable, but it would still be inconsistent to call the variable `importAssertions`, I don't really mind either way.

GeoffreyBooth replied:
What would be different here if you weren't following the new spec?

guybedford commented on lib/internal/modules/esm/module_map.js:
```diff
@@ -1,17 +1,119 @@
+   * @param {Record<string,string>} importAssertions
+   * @returns {{
+   *   serializedModuleRequest: string,
+   *   internalCache: Record<string, Promise<import('./loader').ModuleExports> | import('./loader').ModuleExports>,
```
It might be worth entirely separating the serialization process from the lookup process?

Something like:

```js
const requestKey = resolveCache.serializeKey(specifier, attributes);
const absoluteResolvedURL = await resolve(specifier, parentURL, attributes);
const resolvedKey = resolveCache.serializeKey(absoluteResolvedURL, attributes);
resolveCache.set(requestKey, parentKey, resolvedKey);
```

Ideally we should also be using `parentKey` not `parentURL` I think. And then note that you probably want to avoid repeat serialization so you want to be maintaining these keys on the module jobs themselves for easy lookup.

Does that make sense? Let me know if I can clarify any of the above.

aduh95 replied:
> Ideally we should also be using `parentKey` not `parentURL` I think

That's a good remark, I didn't consider this. I don't think we have access to the `parentKey` when doing dynamic imports, do we? 🤔 Using `parentURL` doesn't go against the spec IIUC, it does add a limitation that's not in the spec (but it likely won't matter that much in practice, I'd expect most use case would be more than fine if modules that have the same URL use the same resolve cache), so I'm tempted to say `parentURL` is good enough. wdyt?

> It might be worth entirely separating the serialization process from the lookup process?

That's probably a good idea. The current implementation is probably not going to stay as Jacob is trying to refactor all that is his PR, so I'm hesitant on spending too much time crafting an implementation that we want to get rid of anyway. What I'm mostly interested in is to get the tests to land, so we can refactor with more confidence.

GeoffreyBooth replied:
What is `parentKey` exactly? A serialized `parentURL` plus assertions/attributes?

guybedford reviewed (commented):
Seems in line with the overall spec approaches for attributes and module maps. That said, it would be nice to try and keep the implementation as simple as possible. I think there's an omission with `parentURL` being used where `parentKey` should be used.

Another approach here might be to separate the import attributes keying work from the dynamic import work, although doing them together seems okay too but most of the work seems to be import attributes here.

Simplifying this PR to just focus on the current URL model without import attributes might be another way to have a more progressive start as well.

aduh95 commented:
I've made a few changes:

- the resolve cache is back at storing the resolve result, rather than the job.
- the cache is used only when there are no loaders involved (there wasn't consensus if we wanted to also have a cache when loaders are involved; the upside is that it simplify the cache setup as the default resolve is sync, so no async failure to take care of).
- I added tests for the `ResolveCache` class.
- I removed the not-exactly-related test change to its own PR.

aduh95 commented:
Benchmark CI: https://ci.nodejs.org/view/Node.js%20benchmark/job/benchmark-node-micro-benchmarks/1351/

aduh95 commented:
The perf improvement is HUGE (147.25% faster than `main`, ±12.59%) when re-importing an existing file without any import attributes, no measurable difference for an import that fails.

aduh95 commented:
Benchmark CI: https://ci.nodejs.org/view/Node.js%20benchmark/job/benchmark-node-micro-benchmarks/1352/

targos commented on lib/internal/modules/esm/module_map.js:
```diff
@@ -1,17 +1,89 @@
+    if (internalCache == null) {
+      super.set(parentURL, internalCache = { __proto__: null });
+    }
+    return serializedKey in internalCache;
```
Is this faster than `ObjectHasOwn` ?

aduh95 replied:
It doesn't really matter, this method is not used anywhere, and only provided for `Map` compatibility.

targos commented on lib/internal/modules/esm/module_map.js:
```diff
@@ -1,17 +1,89 @@
+   * @returns {import('./loader').ModuleExports | Promise<import('./loader').ModuleExports>}
+   */
+  get(serializedKey, parentURL) {
+    let internalCache = super.get(parentURL);
+    if (internalCache == null) {
+      super.set(parentURL, internalCache = { __proto__: null });
+    }
```
Have you tried to refactor this into an private method?

ljharb commented on lib/internal/modules/esm/module_job.js:
```diff
@@ -140,7 +140,7 @@ class ModuleJob {
           e.message);
         const { url: childFileURL } = await this.loader.resolve(
-          childSpecifier, parentFileUrl,
+          childSpecifier, parentFileUrl, kEmptyObject,
```
nit, to avoid future diff churn:
```suggestion
          childSpecifier,
          parentFileUrl,
          kEmptyObject,
```

ljharb commented on lib/internal/modules/esm/module_map.js:
```diff
@@ -1,17 +1,85 @@
+    // as identical.
+    return specifier + '::' + ArrayPrototypeJoin(
+      ArrayPrototypeMap(
+        ArrayPrototypeSort(ObjectKeys(importAssertions)),
+        (key) => JSONStringify(key) + JSONStringify(importAssertions[key])),
```
```suggestion
        ArrayPrototypeSort(ObjectEntries(importAssertions)),
        (key, value) => JSONStringify(key) + JSONStringify(value)),
```
also, if you don't have a delimiter, then can't someone have an object like `{ a: 'bc', ab: 'c' }`?

aduh95 replied:
`%Array.prototype.sort%` stringily all members, which in this case are arrays, which would be susceptible to prototype mutation – and also maybe not spec compliant, the spec only want to sort the keys.

> it seems weird to me to not have a delimiter

Yeah but also this value is not observable, adding a delimiter seems hardly worth the added compute time (however small it is).

ljharb replied:
ahh right, good point, please resolve this

GeoffreyBooth commented on lib/internal/modules/esm/loader.js:
```diff
@@ -20,9 +20,14 @@ const {
-function newModuleMap() {
-  const ModuleMap = require('internal/modules/esm/module_map');
-  return new ModuleMap();
+function newLoadCache() {
+  const { LoadCache } = require('internal/modules/esm/module_map');
+  return new LoadCache();
+}
+
+function newResolveCache() {
+  const { ResolveCache } = require('internal/modules/esm/module_map');
+  return new ResolveCache();
 }
```
Since `resolve` happens first, I feel like we should define its cache first.

```suggestion
function newResolveCache() {
  const { ResolveCache } = require(‘internal/modules/esm/module_map’);
  return new ResolveCache();
}

function newLoadCache() {
  const { LoadCache } = require(‘internal/modules/esm/module_map’);
  return new LoadCache();
}
```

GeoffreyBooth commented on lib/internal/modules/esm/loader.js:
```diff
@@ -74,7 +79,12 @@ class ModuleLoader {
   /**
    * Registry of loaded modules, akin to `require.cache`
    */
-  moduleMap = newModuleMap();
+  loadCache = newLoadCache();
+
+  /**
+   * Registry of resolved URLs
+   */
+  #resolveCache = newResolveCache();
```
Same here, I feel like `resolve` should come first.

aduh95 replied:
```suggestion
  #resolveCache = newResolveCache();

  /**
   * Registry of loaded modules, akin to `require.cache`
   */
  loadCache = newLoadCache();
```

GeoffreyBooth commented on lib/internal/modules/esm/loader.js:
```diff
@@ -315,13 +325,20 @@ class ModuleLoader {
    * @param {ImportAssertions} importAssertions Assertions from the import
    *                                            statement or expression.
-   * @returns {Promise<{ format: string, url: URL['href'] }>}
+   * @returns {{ format: string, url: URL['href'] }}
```
Is this just an unrelated cleanup? (I think it’s fine if it is, but just making sure it’s not something that this PR is changing.)

GeoffreyBooth commented on lib/internal/modules/esm/module_map.js:
```diff
@@ -1,17 +1,85 @@
+ * Future resolutions of the same input (specifier, parent URL and import assertions)
+ * must return the same result if the first attempt was successful, per
+ * https://tc39.es/ecma262/#sec-HostLoadImportedModule.
+ */
```
```suggestion
 * This cache is *not* used when custom loaders are registered.
 */
```

GeoffreyBooth commented on lib/internal/modules/esm/module_map.js:
```diff
@@ -1,17 +1,85 @@
+   * @returns {string}
+   */
+  serializeKey(specifier, importAssertions) {
+    // To serialize the ModuleRequest (specifier + list of import attributes),
+    // we need to sort the attributes by key, then stringifying,
+    // so that different import statements with the same attributes are always treated
```
Until the PR that replaces import assertions with attributes.

```suggestion
    // To serialize the ModuleRequest (specifier + list of import assertions),
    // we need to sort the assertions by key, then stringifying,
    // so that different import statements with the same assertions are always treated
```

GeoffreyBooth commented on lib/internal/modules/esm/module_map.js:
```diff
@@ -1,17 +1,85 @@
+    // we need to sort the attributes by key, then stringifying,
+    // so that different import statements with the same attributes are always treated
+    // as identical.
+    return specifier + '::' + ArrayPrototypeJoin(
+      ArrayPrototypeMap(
+        ArrayPrototypeSort(ObjectKeys(importAssertions)),
+        (key) => JSONStringify(key) + JSONStringify(importAssertions[key])),
+      ',');
```
Is this faster?

```suggestion
    const keys = ObjectKeys(importAssertions);
    
    if (keys.length === 0) {
      return specifier;
    }

    return specifier + '::' + ArrayPrototypeJoin(
      ArrayPrototypeMap(
        ArrayPrototypeSort(keys),
        (key) => JSONStringify(key) + JSONStringify(importAssertions[key])),
      ',');
```

aduh95 replied:
I'm not sure, it would need to be benchmarked, but more importantly it's making the test fail, so I suggest we leave that to a later PR.

GeoffreyBooth replied:
You need to update the test so that imports without assertions don’t get a key of `specifier::` but just `specifier`. That was an intentional change in this code, to skip the cost of concatenation and generating a new string.

aduh95 replied:
Removing `::` puts us at risk of cache collision, updating the test would not be the solution here, it would hide the fact the implementation is not safe.

GeoffreyBooth replied:
> Removing `::` puts us at risk of cache collision

How?

aduh95 replied:
Well see the test case, I've added it because otherwise it's trivial to create a cache collision.

GeoffreyBooth replied:
> Well see the test case, I’ve added it because otherwise it’s trivial to create a cache collision.

Because there might be a `::` within the specifier? If that’s a concern, then just add back the concatenation:

```suggestion
    const keys = ObjectKeys(importAssertions);
    
    if (keys.length === 0) {
      return specifier + '::';
    }

    return specifier + '::' + ArrayPrototypeJoin(
      ArrayPrototypeMap(
        ArrayPrototypeSort(keys),
        (key) => JSONStringify(key) + JSONStringify(importAssertions[key])),
      ',');
```

I would still think that this should be faster than the existing code, as it bypasses the sort and map when there are no assertions.

You might want to consider something other than `::` as the delimiter though, if this is the risk, so that we don’t need to add an unnecessary delimiter for the common case. Perhaps a character that’s disallowed in specifiers, like `\t` or `\n` or something? (Not sure if those are disallowed or what characters are disallowed.)

aduh95 replied:
AFAIK any character is allowed in a specifier string. This does feel like premature optimisation, the benchmark didn't show any significant perf regression with the current implementation. Tests are passing, I'd prefer to land that now, and we can do follow-up PRs to optimize further if necessary.

GeoffreyBooth replied:
I disagree, I feel like there's no way this isn't faster and I've said for a while now that we should optimize for the common case of no assertions. I'm happy to rerun CI for you to get it green again.

aduh95 replied:
If we land as is, and open a follow up PR, it will be trivial to compute if your proposal is indeed better for perf and how much.

targos replied:
Even if it doesn't improve the perf noticeably, I like @GeoffreyBooth's suggestion because it makes the code easier to understand. It becomes obvious that the complex part is here to deal with import assertions.

GeoffreyBooth commented on lib/internal/modules/esm/module_map.js:
```diff
@@ -1,17 +1,85 @@
+      ',');
+  }
+
+  #getInternalCache(parentURL) {
```
I feel like this could get a better name. Maybe this? Open to other suggestions.
```suggestion
  #getModuleCachedImports(parentURL) {
```

Can this also please get a JSDoc or simple comment explaining why we have this cache within a cache?

GeoffreyBooth commented on lib/internal/modules/esm/module_map.js:
```diff
@@ -1,17 +1,85 @@
+  #getInternalCache(parentURL) {
+    let internalCache = super.get(parentURL);
+    if (internalCache == null) {
+      super.set(parentURL, internalCache = { __proto__: null });
```
```suggestion
      super.set(parentURL, internalCache = new SafeMap());
```

And associated changes.

aduh95 replied:
I'm not going to do that, if `Map` is faster than an object, we should update `LoadCache` as well and that's out of scope for this PR.

GeoffreyBooth replied:
Fair enough. @JakobJingleheimer if you still do a refactor after this, maybe add this to the list of improvements?

GeoffreyBooth reviewed (commented):
Thank you, this makes much more sense to me. It’s actually quite straightforward to follow. Excellent work!

Does the resolve cache having a value of the resolve output, as opposed to the load output, decrease performance noticeably? Because if so, I can sullenly acknowledge that the unintuitive map of resolve input to load output would be a better choice, as performance probably trumps readability (though we should at least have a detailed comment explaining the choice). But if what we have here is faster or equivalent, by all means let’s keep this.

aduh95 commented on lib/internal/modules/esm/loader.js:
```diff
@@ -74,7 +79,12 @@ class ModuleLoader {
   /**
    * Registry of loaded modules, akin to `require.cache`
```
```suggestion
   * Registry of resolved specifiers
```

GeoffreyBooth reviewed (approved):
Thank you!
